pkg/controller/crdmirroring/crdmirroring_controller.go

```diff
@@ -0,0 +1,209 @@
+	c.listerSycned = c.informer.Informer().HasSynced
+	c.legacyInformer.Informer().AddEventHandler(legacyHandlers)
+	c.legacyListerSynced = c.legacyInformer.Informer().HasSynced
+	c.mirroringHandler = crdhandler.NewMirroringHandler(c.informer, c.legacyInformer, client, legacyClient, c.CRDName)
+
+	return c
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %vMirroringController", c.CRDName)
+	defer klog.Infof("Shutting down %vMirroringController", c.CRDName)
+
+	if !cache.WaitForNamedCacheSync(fmt.Sprintf("%vMirroringController", c.CRDName), stopCh, c.listerSycned, c.legacyListerSynced) {
+		return
+	}
+
+	klog.Infof("Starting %d worker threads", defaultWorkers)
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, c.workerLoopPeriod, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	cKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(cKey)
+
+	err := c.syncMirroring(cKey.(string))
+	c.handleErr(err, cKey)
+
+	return true
+}
+
+func (c *Controller) syncMirroring(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing for %q legacy CRD. (%v)", key, time.Since(startTime))
+	}()
+
+	klog.V(4).Infof("Sync mirroring CRD (%q)", key)
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	// Get he legacy object, and if got an error that is not IsNotFound, return the error
+	legacyObj, err := c.mirroringHandler.GetLegacyObject(namespace, name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// Get he legacy object, and if got an error that is not IsNotFound, return the error
+	newObj, err := c.mirroringHandler.GetNewObject(namespace, name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// If the legacy object doesn't exist and the new object exists as well as the new object has "ManagedBy" annotation,
+	// delete the mirroring new object.
+	if legacyObj == nil && newObj != nil {
+		_, exist := newObj.GetAnnotations()[types.ManagedBy]
+		if exist {
+			klog.Infof("Legacy %s/%s object not found, deleting the mirroring new object", namespace, name)
+			return c.mirroringHandler.DeleteNewObject(namespace, name)
+		}
+	}
+
+	// If the legacy object exists and the new object exists as well as the legacy object doesn't have "crd.antrea.io/stop-mirror"
+	// annotation, creating a new object.
+	if legacyObj != nil && newObj == nil {
+		_, exist := legacyObj.GetAnnotations()[types.MirroringStatus]
+		if !exist {
+			klog.Infof("New %s/%s object not found, mirroring a new object", namespace, name)
+			return c.mirroringHandler.AddNewObject(legacyObj)
+		}
+	}
+
+	// If both the legacy object and the new exist, do something according their annotation.
+	if legacyObj != nil && newObj != nil {
```

Comment:
This check could be in the begining so the 4 cases (both missing, both present, one missing and one present...) are organized more logcially and terminating the process earlier:
```
if legacyObj == nil && newObj == nil {
    return
}

if legacyObj != nil && newObj == nil {
    ....
    return
}

if legacyObj == nil && newObj != nil {
    ....
    return
}

// Both present case
....
```

Reply:
Corrected as your suggestion.